### ISSUE TITLE
Add local files (PDFs) as bookmarks

### ISSUE THREAD

**Issue by antintin0** [zombieFox/nightTab#313]
**What kind of feature request is this?**

_Feature improvement
---

**Is your feature request related to a problem? Please describe.**

It's annoying to not be able to add some PDFs as bookmarks

**Describe the solution you'd like**

Allow adding files like PDFs as bookmarks
Nighttab seems to only allow https:// and not file:/// 

---

**Describe alternatives you've considered**
NA

---

**Comment by zombieFox** [comment]
Hi, this has been requested before [(1)](https://github.com/zombieFox/nightTab/issues/187) [(2)](https://github.com/zombieFox/nightTab/issues/152). Chrome does not allow extensions to access privileged URLs. Please see [Protected URLs](https://github.com/zombieFox/nightTab/wiki/Protected-URLs) for details.

**Comment by antintin0** [comment]
I understand it's best to keep permissions at a minimum or request none at all, but wouldn't it make sense to add a simple switch in settings for an optional storage permission? I tried to add the "storage" permission into the manifest, and that made the "allow access to file URLs" switch appear in the chrome extension management next to "allow in incognito," but it still wouldn't open up file URLs

**Comment by antintin0** [comment]
> Hi, this has been requested before [(1)](https://github.com/zombieFox/nightTab/issues/187) [(2)](https://github.com/zombieFox/nightTab/issues/152). Chrome does not allow extensions to access privileged URLs. Please see [Protected URLs](https://github.com/zombieFox/nightTab/wiki/Protected-URLs) for details.

I understand it's best to keep permissions at a minimum or request none at all, but wouldn't it make sense to add a simple switch in settings for an optional storage permission? I tried to add the "storage" permission into the manifest, and that made the "allow access to file URLs" switch appear in the chrome extension management next to "allow in incognito," but it still wouldn't open up file URLs

**Comment by zombieFox** [comment]
> but it still wouldn't open up file URLs

Totally understand that progressive permissions is a good approach, it's something I'll be looking into going forward. However I don't believe even with permissions Chrome will allow a link to be created to a local file. 

nightTab, at a basic level, is essentially making links (`<a href="..."></a>`) to URLs. And after experiments and testing I have not been able to create a link which can access a users system file structure. That also just sounds totally exploitable and dangerous. I do not believe what you ware asking for will be possible due to security restrictions. 

tl;dr: "allow access to file URLs" might allow an extension to read the files on the system but, as far as I can see, not link to the files.

**Comment by metruzanca** [comment]
I've found this working implementation for chrome but theres nothing for firefox that I could find.
https://github.com/tksugimoto/chrome-extension_open-local-file-link/blob/master/chrome-extension/content_script.js

It also seems to not use special permissions.

Firefox doesn't seem to easily allow this sort of thing.

**Comment by zombieFox** [comment]
> I've found this working implementation for chrome but theres nothing for firefox that I could find.
https://github.com/tksugimoto/chrome-extension_open-local-file-link/blob/master/chrome-extension/content_script.js

Interesting, a bit hard to follow what is happening in that function -- I'll see if I can test it out locally when I find some time. 👍🏽

**Comment by metruzanca** [comment]
> Interesting, a bit hard to follow what is happening in that function -- I'll see if I can test it out locally when I find some time. 👍🏽

yeah, its a bit wack. The while loop is for drilling down to the a tag since thats a document event and its listening on window click (super inefficient)

The important part is just 
```js
chrome.runtime.sendMessage({
	method: 'openLocalFile',
	localFileUrl: url,
});
```
which unfortunately is not something thats currently documented on chrome's docs. Maybe its in some example from their github.

> Side note: I would really like to maintain feature parity between Firefox and Chrome.

I'd would agree here as it would only add to the confusion and this current way is pretty hacky.

**Comment by antintin0** [comment]
> > but it still wouldn't open up file URLs
> 
> Totally understand that progressive permissions is a good approach, it's something I'll be looking into going forward. However I don't believe, even with the right permissions Chrome, will allow a link to be created to a local file.
> 
> nightTab, at a basic level, is essentially making links (`<a href="..."></a>`) to URLs. And after experiments and testing I have not been able to create a link which can access a users system file structure. That also just sounds totally exploitable and dangerous. I do not believe what you are asking for will be possible due to security restrictions.
> 
> tl;dr: "allow access to file URLs" might allow an extension to read the files on the system but, as far as I can see, not link to the files.

I'm not sure what you mean by not being able to link to files through URLs because I can bookmark a PDF as a file:///[path] to chrome's built in bookmark system, and copy pasting that file:///[path] URL  correctly opens the PDF just like any https:// link

**Comment by metruzanca** [comment]
It is indeed a limitation of modern browsers as giving websites access to the file system would be a major security risk. `chrome-extension://` and `moz-extension://` share the same limitations as `https-served` content to use `file://` protocol. However, "websites" can use application protocols which are handled differently.

After a quick search I found this which has some further reading: [navigating to file urls](https://textslashplain.com/2019/10/09/navigating-to-file-urls/).

Bookmarks in chrome/firefox can use `file://` URLs because they are not vulnerable to the same exploits as `https-served` content. (which all stem from the fact that theres no guarantee that `https-served` content isn't doing something malicious)

**Comment by antintin0** [comment]
> I've found this working implementation for chrome but theres nothing for firefox that I could find. https://github.com/tksugimoto/chrome-extension_open-local-file-link/blob/master/chrome-extension/content_script.js
> 
> It also seems to not use special permissions.
> 
> Firefox doesn't seem to easily allow this sort of thing.

So this extension allows the https:// content to switch over to file:/// content? Is there anyway to bypass the restriction for chrome-extension://?
